compiler/rustc_mir_transform/src/inline.rs

```diff
@@ -102,46 +103,80 @@ fn inline<'tcx>(tcx: TyCtxt<'tcx>, body: &mut Body<'tcx>) -> bool {
         tcx,
         param_env,
         codegen_fn_attrs,
-        history: Vec::new(),
-        changed: false,
+        queue: BinaryHeap::new(),
         caller_is_inline_forwarder: matches!(
             codegen_fn_attrs.inline,
             InlineAttr::Hint | InlineAttr::Always
         ) && body_is_forwarder(body),
     };
-    let blocks = START_BLOCK..body.basic_blocks.next_index();
-    this.process_blocks(body, blocks);
-    this.changed
+    let initial_blocks = START_BLOCK..body.basic_blocks.next_index();
+    this.enqueue_new_candidates(body, initial_blocks);
+
+    let mut changed = false;
+    let mut remaining_cost = tcx.sess.opts.unstable_opts.inline_mir_total_threshold.unwrap_or(300);
+    let mut remaining_count = MAX_INLINED_CALLEES_PER_BODY;
```

Comment:
Shouldn't this go by some size heuristic? if the callee is absolutely tiny (e.g. just another call) then I'd expect the cost of inlining to be ~zero (replacing a call with another call), which means we can inline an unlimited amount of those cases.

Reply:
Ah, looking at the testcases I guess the scopes make this non-zero cost?